### PR TITLE
refactor(sumologicextension): refactor registration and credentials storage calls

### PR DESCRIPTION
**Pull request by pmalek-sumo**
Related to discussion in: https://github.com/SumoLogic/sumologic-otel-collector/pull/404#discussion_r801525739

### CODE REVIEW

File: pkg/extension/sumologicextension/extension.go [hunk 1]
```diff
@@ -200,104 +168,146 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
+// getLocalCredentials returns the credentials either retrieved from local credentials
+// storage in cases they were available there or it registers the collector and
+// returns the credentials obtained from the API.
+// It also returns a boolean flag indicating if the successful registration
+// with the API took place and an error.
+func (se *SumologicExtension) getLocalCredentials(ctx context.Context) (credentials.CollectorCredentials, error) {
+	if !se.credentialsStore.Check(se.hashKey) {
+		return credentials.CollectorCredentials{},
+			fmt.Errorf("collector credentials not found (hash key: %s)", se.hashKey)
+	}
```
Comment (swiatekm-sumo) [inline pkg/extension/sumologicextension/extension.go:298]: Is there anything this check gives us that the `.Get` below doesn't?

File: pkg/receiver/telegrafreceiver/converter.go [hunk 2]
```diff
@@ -42,6 +42,10 @@ func newConverter(separateField bool, logger *zap.Logger) MetricConverter {
 	}
 }
 
+var _filter = map[string]struct{}{
+	"http_listener_v2_path": {},
+}
+
 // Convert converts telegraf.Metric to pdata.Metrics.
```
Comment (swiatekm-sumo) [inline pkg/receiver/telegrafreceiver/converter.go:49]: Why are these changes here?

**Review by swiatekm-sumo — CHANGES_REQUESTED** [review]

File: pkg/extension/sumologicextension/extension.go [hunk 3]
```diff
@@ -200,104 +168,146 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
 
+func (se *SumologicExtension) validateCredentials(
+	ctx context.Context,
+) error {
+	se.logger.Info("Validating collector credentials...",
+		zap.String(collectorIdField, se.registrationInfo.CollectorCredentialId),
+		zap.String(collectorNameField, se.registrationInfo.CollectorName),
+	)
+
+	return se.sendHeartbeat(ctx)
```
[the remark this answers is not in the document]
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:190]: Well we could but this really checks how the credentials were injected into the http client's transport 🤔 So however we do this it will be implicit anyway (so some extent).

I'm open to suggestions.

File: pkg/extension/sumologicextension/extension.go [hunk 4]
```diff
@@ -200,104 +168,146 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
-	if se.credentialsStore.Check(se.hashKey) {
-		colCreds, err = se.credentialsStore.Get(se.hashKey)
-		if err != nil {
-			return credentials.CollectorCredentials{}, false, err
-		}
-		se.collectorName = colCreds.CollectorName
-		if !se.conf.Clobber {
-			if colCreds.ApiBaseUrl != "" {
-				se.baseUrl = colCreds.ApiBaseUrl
+	if !se.conf.Clobber {
```
[the remark this answers is not in the document]
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:229]: this is going away anyway as we'll be adding a separate flag for re-registration

File: pkg/extension/sumologicextension/extension.go [hunk 5]
```diff
@@ -200,104 +168,146 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
-	return colCreds, registrationDone, err
+	colCreds, err = se.getCredentialsByRegistering(ctx)
+	if err != nil {
+		return credentials.CollectorCredentials{}, err
+	}
+
+	return colCreds, nil
+}
+
+func (se *SumologicExtension) getCredentialsByRegistering(ctx context.Context) (credentials.CollectorCredentials, error) {
```
[the remark this answers is not in the document]
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:272]: Seems consistent to `getLocalCredentials` though 🤔

File: pkg/receiver/telegrafreceiver/converter.go (hunk 2 above)
Comment (pmalek-sumo) [reply pkg/receiver/telegrafreceiver/converter.go:49]: Derp, will fix this

File: pkg/extension/sumologicextension/extension.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/extension/sumologicextension/extension.go:272]: I guess it's fine, it captures the meaning well enough.

File: pkg/extension/sumologicextension/extension.go (hunk 1 above)
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go]: An explicit check only. Changed to `.Get`

File: pkg/extension/sumologicextension/extension.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (swiatekm-sumo) [reply pkg/extension/sumologicextension/extension.go:190]: I think what it *should* check is whether the generated transport lets us do a valid request. Something like:
```go
httpClient := se.getHttpClient(creds)
se.sendHeartbeatWithHttpClient(httpClient)
```
and use these functions for normal heartbeats as well.

File: pkg/extension/sumologicextension/extension.go [hunk 6]
```diff
@@ -200,104 +169,166 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
-		se.collectorName = colCreds.CollectorName
-		if !se.conf.Clobber {
-			if colCreds.ApiBaseUrl != "" {
-				se.baseUrl = colCreds.ApiBaseUrl
+	if !se.conf.Clobber {
+		colCreds, err = se.getLocalCredentials(ctx)
+		if err == nil {
+			if errI := se.injectCredentials(colCreds, host); errI != nil {
+				return credentials.CollectorCredentials{}, errI
+			}
```
Comment (swiatekm-sumo) [inline pkg/extension/sumologicextension/extension.go:258]: With the changes to `.validateCredentials`, can we now move this call to after the credentials were validated?

Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go]: What we could do is we could remove `injectCredentials` call and move it to `validateCredentials` as we'll need to set the `.registrationInfo` field anyway.

Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go]: I've moved stuff around PTAL @swiatekm-sumo

File: pkg/extension/sumologicextension/extension.go [hunk 7]
```diff
@@ -200,104 +169,161 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
+	host component.Host,
+	colCreds credentials.CollectorCredentials,
+) error {
+	se.logger.Info("Validating collector credentials...",
+		zap.String(collectorIdField, se.registrationInfo.CollectorCredentialId),
+		zap.String(collectorNameField, se.registrationInfo.CollectorName),
+	)
+
+	if err := se.injectCredentials(colCreds, host); err != nil {
+		return err
```
Comment (swiatekm-sumo) [inline pkg/extension/sumologicextension/extension.go:194]: Do we actually need this here? It doesn't look like it does anything other than setting the client, which we're explicitly using for the heartbeat anyway.

File: pkg/extension/sumologicextension/extension.go [hunk 8]
```diff
@@ -200,104 +169,161 @@ func (se *SumologicExtension) Start(ctx context.Context, host component.Host) er
...
+}
+
+func (se *SumologicExtension) validateCredentials(
+	ctx context.Context,
+	host component.Host,
+	colCreds credentials.CollectorCredentials,
+) error {
+	se.logger.Info("Validating collector credentials...",
+		zap.String(collectorIdField, se.registrationInfo.CollectorCredentialId),
+		zap.String(collectorNameField, se.registrationInfo.CollectorName),
```
Comment (swiatekm-sumo) [inline pkg/extension/sumologicextension/extension.go:190]: I think these should use `colCreds` instead.

File: pkg/extension/sumologicextension/extension.go (hunk 7 above)
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:194]: It is needed as it sets the credentials in `registrationInfo` and the client (as you mentioned).

File: pkg/extension/sumologicextension/extension.go (hunk 8 above)
Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:190]: changed

**Review by swiatekm-sumo — APPROVED** [review]
